torch/csrc/jit/tensorexpr/schedule.cpp

```diff
@@ -399,6 +399,126 @@ class FunctionInliner : public IRMutator {
   std::unordered_set<const Var*> func_var_set_;
 };
 
+/**
+ * Actual algorithm for "inlining" of rand().  Bind variables to rand() calls.
+ */
+class RandomInliner : public IRMutator {
+ public:
+  RandomInliner(const std::unordered_set<Function*>& funcs) : funcs_(funcs) {
+    for (auto const* func : funcs_) {
+      func_vars_.insert(func->func_var(0));
+    }
+  }
+
+  static bool is_inner_loop(const For* loop) {
+    Block* body = dynamic_cast<Block*>(loop->body());
+    if (!body) {
+      return false;
+    }
+    if (body->nstmts() != 1) {
+      return false;
+    }
+    return dynamic_cast<For*>(*body->stmts().begin()) == nullptr;
+  }
+
+  Stmt* mutate(const For* v) override {
+    const Var* var = v->var();
+    const Expr* start = v->start();
+    const Expr* stop = v->stop();
+    Stmt* body = v->body();
+    LoopOptions loop_options = v->loop_options();
+
+    // Put the new "let"s in the innermost loop.
+    if (!is_inner_loop(v)) {
+      Stmt* new_body = body->accept_mutator(this);
+      if (new_body == body) {
+        return const_cast<For*>(v);
+      }
+      if (new_body == nullptr) {
+        return nullptr;
+      }
+      return new For(var, start, stop, new_body, loop_options);
+    }
+
+    Stmt* new_body = body->accept_mutator(this);
+    for (auto const& p : random_vars_) {
+      Var* v = p.second;
+      new_body = new LetStmt(v, new Intrinsics(kRand, v->dtype()), new_body);
```

Comment:
It would be helpful to add a comment describing how we're inserting the let statement and why this helps (maybe as a class comment).